### PR TITLE
feat: funds distribution

### LINKED ISSUE #243: ✨ Distribution Messages
## ✨ Feature

Distribution messages so funds can be burnt/refunded/sent.

## Motivation

We need to pay out, burn, and return the funds depending on the situation.

## Implementation

See notion.

### PR DESCRIPTION
**Pull request by gluax**
## Motivation

Distribution of data request escrow funds.

## Explanation of Changes

- Iterate over provided data request ids in removal.
- Iterate over each message for the data request.
  - It emits events for each message
  - It moves appropriate funds
- If funds are leftover uses the refund, and refund message.
  - The reason there is a refund_type is so for events if we are doing a full refund we can give a reason. But most times it will be remainder payback if it exists.
- Removes the Escrow from the map.

## Testing

- Test util changes better track executor funds.
- Test that messages work and funds update as expected

## Related PRs and Issues

Waiting for https://github.com/sedaprotocol/seda-common-rs/pull/34 to be merged so we can update common to main.
Closes #243

### CODE REVIEW

**Review by Thomasvdam — COMMENTED** [review]
With our current design this won't work since we don't register a SEDA address for an executor node.

I think we should differentiate between data proxy payouts and executor payouts. Data proxy is a simple bank transfer, but an executor payout is an internal accounting change on the contract with the funds staying on the contract.
We discussed having moving the funds to the identities' staking pool, and anything that exceeds the minimum_to_stake amount goes straight to the to_withdraw pool. Given that we're starting with an allowlisted pool of executors we could opt to simplify this and always move it to one of them.

File: contract/src/msgs/data_requests/sudo/remove_requests.rs [hunk 1]
```diff
@@ -1,14 +1,112 @@
...
+    token: &str,
+) -> Result<(Event, Vec<BankMsg>), ContractError> {
+    // find the data request from the committed pool (if it exists, otherwise error)
+    let dr_id = Hash::from_hex_str(&dr_id_str)?;
+    state::load_request(deps.storage, &dr_id)?;
+
+    let block_height: u64 = env.block.height;
+
+    let mut event =
+        Event::new("remove-dr").add_attributes([("dr_id", dr_id_str), ("block_height", block_height.to_string())]);
```
Comment (Thomasvdam) [inline contract/src/msgs/data_requests/sudo/remove_requests.rs:37]: Can we make the event `seda-remove-dr`? 🙏

**Comment by gluax** [comment]
> With our current design this won't work since we don't register a SEDA address for an executor node.
> 
> I think we should differentiate between data proxy payouts and executor payouts. Data proxy is a simple bank transfer, but an executor payout is an internal accounting change on the contract with the funds staying on the contract. We discussed having moving the funds to the identities' staking pool, and anything that exceeds the minimum_to_stake amount goes straight to the to_withdraw pool. Given that we're starting with an allowlisted pool of executors we could opt to simplify this and always move it to one of them.

Oh I totally forgot about that it's been so long since we talked about it.

So need to add another `DistributionMessage` that is a executor reward that takes the identity and moves it to their staking pool up to the minimum amount to stake. The rest gets point into a new `RewardEscrow` the identity could withdraw from? So I need to add a new execute method to withdraw from that, and probably a query function to view your reward escrow?

**Comment by Thomasvdam** [comment]
Right now an executor has `stake` and `funds_ready_to_withdraw` or something similar. I guess we can use those two fields for now and don't bother with a unstaking period.

File: contract/Cargo.toml
Comment (mariocao) [inline contract/Cargo.toml:1]: Do we want to bump the version here too? Or not yet?

Comment (gluax) [reply contract/Cargo.toml:1]: I forgot I'll do that now ^^

Comment (gluax) [reply contract/Cargo.toml:1]: done in #242.

File: contract/src/msgs/data_requests/sudo/remove_requests.rs [hunk 2]
```diff
@@ -1,14 +1,147 @@
...
+                        "kind": to_json_string(&message.kind)?,
+                    }))?,
+                );
+            }
+        }
+    }
+
+    if !dr_escrow.amount.is_zero() {
+        bank_messages.push(BankMsg::Send {
+            to_address: dr_escrow.staker.to_string(),
```
Comment (mariocao) [inline contract/src/msgs/data_requests/sudo/remove_requests.rs:115]: This should be `poster`
```suggestion
            to_address: dr_escrow.poster.to_string(),
```

**Review by mariocao — APPROVED** [review]

**Pull request merged**